File: Training/Advanced_Cluster_Usage/Exercise2/R/mp_mem_test.R
library(parallel)

# Parallel function to create a symmetric random matrix
create_symmetric_matrix_parallel <- function(n) {
  matrix <- matrix(0.0, nrow = n, ncol = n)

  # Set number of threads
  cores <- 2  # explicitly set number of threads to 2
  cat("Using", cores, "cores for parallel execution\n")

  # Split the indices for parallel execution
  indices <- split(1:n, cut(1:n, cores, labels = FALSE))

  # Define parallel task
  parallel_task <- function(rows) {
    partial_matrix <- matrix(0.0, nrow = length(rows), ncol = n)
    for (idx in seq_along(rows)) {
      i <- rows[idx]
      for (j in 1:i) {
        value <- runif(1)
        partial_matrix[idx, j] <- value
        partial_matrix[idx, j] <- value
      }
    }
    return(partial_matrix)
  }

  # Execute in parallel
  results <- mclapply(indices, parallel_task, mc.cores = cores)

  # Combine results into the matrix
  current_row <- 1
  for (partial_result in results) {
    rows_count <- nrow(partial_result)
    matrix[current_row:(current_row + rows_count - 1), ] <- partial_result
    current_row <- current_row + rows_count
  }

  # Mirror lower triangular to upper triangular part
  matrix[upper.tri(matrix)] <- t(matrix)[upper.tri(matrix)]

  return(matrix)
}

# Main function
main_parallel <- function() {
  n <- 20000
  cat("Creating a symmetric random matrix of size", n, "x", n, "in parallel...\n")

  h <- create_symmetric_matrix_parallel(n)

  cat("Hamiltonian matrix created successfully!\n")

  cat("Top-left 5x5 corner of the matrix:\n")
  print(h[1:5, 1:5])
}

# Run the parallel main function
main_parallel()

File: Training/Advanced_Cluster_Usage/Exercise2/Python/mp_mem_test.py
import random
import threading

def create_symmetric_matrix_parallel(n, num_threads=2):
    """
    Creates a symmetric random matrix of size n x n using Python threading.
    """
    matrix = [[0.0 for _ in range(n)] for _ in range(n)]

    def worker(start_row, end_row):
        for i in range(start_row, end_row):
            for j in range(i + 1):
                value = random.random()
                matrix[i][j] = value
                matrix[j][i] = value

    # Calculate rows for each thread
    rows_per_thread = n // num_threads
    threads = []

    for i in range(num_threads):
        start_row = i * rows_per_thread
        # Ensure the last thread covers any remaining rows
        end_row = (start_row + rows_per_thread) if i < num_threads - 1 else n
        thread = threading.Thread(target=worker, args=(start_row, end_row))
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()

    return matrix

def main():
    n = 30000  # Matrix dimension

    print(f"Creating a symmetric random matrix of size {n}x{n} using 2 threads...")

    h = create_symmetric_matrix_parallel(n, num_threads=2)

    print("Hamiltonian matrix created successfully!")

    # Print a small portion of the matrix (optional)
    print("Top-left 5x5 corner of the matrix:")
    for row in h[:5]:
        print(row[:5])

if __name__ == "__main__":
    main()

